src/ReservoirPriceOracle.sol

```diff
@@ -350,43 +483,89 @@ contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.
     /// @param aToken0 Address of the lower token
     /// @param aToken1 Address of the higher token
     /// @param aRoute Path with which the price between aToken0 and aToken1 should be derived
-    // should we make this recursive, meaning for a route A-B-C-D
-    // besides defining A-D, A-B, B-C, and C-D, we also define
-    // A-B-C, B-C-D ?
-    function setRoute(address aToken0, address aToken1, address[] calldata aRoute) external onlyOwner {
+    function setRoute(address aToken0, address aToken1, address[] memory aRoute) public onlyOwner {
+        uint256 lRouteLength = aRoute.length;
+
         if (aToken0 == aToken1) revert RPC_SAME_TOKEN();
         if (aToken1 < aToken0) revert RPC_TOKENS_UNSORTED();
-        if (aRoute.length > MAX_ROUTE_LENGTH || aRoute.length < 2) revert RPC_INVALID_ROUTE_LENGTH();
-        if (aRoute[0] != aToken0 || aRoute[aRoute.length - 1] != aToken1) revert RPC_INVALID_ROUTE();
+        if (lRouteLength > MAX_ROUTE_LENGTH || lRouteLength < 2) revert RPC_INVALID_ROUTE_LENGTH();
+        if (aRoute[0] != aToken0 || aRoute[lRouteLength - 1] != aToken1) revert RPC_INVALID_ROUTE();
 
-        _route[aToken0][aToken1] = aRoute;
-        emit Route(aToken0, aToken1, aRoute);
+        bytes32 lSlot = aToken0.calculateSlot(aToken1);
+
+        // simple route
+        if (lRouteLength == 2) {
+            uint256 lToken0Decimals = IERC20(aToken0).decimals();
+            uint256 lToken1Decimals = IERC20(aToken1).decimals();
+            if (lToken0Decimals > 18 || lToken1Decimals > 18) revert RPC_UNSUPPORTED_TOKEN_DECIMALS();
+
+            int256 lDiff = int256(lToken1Decimals) - int256(lToken0Decimals);
+
+            bytes32 lData = FlagsLib.FLAG_SIMPLE_PRICE.combine(lDiff);
+            assembly {
+                // Write data to storage.
+                sstore(lSlot, lData)
+            }
+        }
+        // composite route
+        else {
+            address lSecondToken = aRoute[1];
+            address lThirdToken = aRoute[2];
+            // Set the uppermost byte of lFirstWord to FLAG_COMPOSITE_NEXT for intermediate hops
+            bytes32 lFirstWord = FlagsLib.FLAG_COMPOSITE_NEXT << 248;
+            // Move the address to start on the 2nd byte.
+            bytes32 lSecondTokenData = bytes32(bytes20(lSecondToken)) >> 8;
+            bytes32 lThirdTokenFirst10Bytes = bytes32(bytes20(lThirdToken)) >> 176;
+
+            // Trim away the first 10 bytes since we only want the last 10 bytes.
```

Comment:
is this comment correct?

Reply:
yes. When we convert address => bytes20 we don't lose any address data. Then when we convert bytes20 => bytes32 it pads right. So when address converted into bytes32 it occupies the top 20 bytes. Shifting by 80 bits will get rid of the top 10 bytes of the address, leaving the bottom 10 bytes of the address on the top 10 bytes of the bytes32